// Copyright 2016, 2017 Intel Corporation
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// -----------------------------------------------------------------------------

syntax = "proto3";
option java_multiple_files = true;
option java_package = "sawtooth.sdk.protobuf";
option go_package = "network_pb2";

// The connect message from a client to the server
message ConnectMessage {
    string identity = 1;
    string endpoint = 2;
}

// The disconnect message from a client to the server
message DisconnectMessage {
    string identity = 1;
}

// The registration request from a peer to the validator
message PeerRegisterRequest {
    string identity = 1;
    string endpoint = 2;
}

// The unregistration request from a peer to the validator
message PeerUnregisterRequest {
    string identity = 1;
}

message GetPeersRequest {
    string identity = 1;
}

message GetPeersResponse {
    repeated string peer_endpoints = 1;
}

message PingRequest {
}

message GossipMessage {
    bytes content = 1;
    string content_type = 2;
}

// A response sent from the validator to the peer acknowledging message
// receipt
message NetworkAcknowledgement {
    enum Status {
        OK = 0;
        ERROR = 1;
    }

    Status status = 1;
}

message GossipBlockRequest {
    // The id of the block that is being requested
    string block_id = 1;

    // The identity of the validator that is requesting the block
    bytes node_id = 2;

    // A random string that provides uniqueness for requests with
    // otherwise identical fields.
    string nonce = 3;

}

message GossipBlockResponse {
    // The block
    bytes content = 1;

    // The identity of the validator that requested the block
    bytes node_id = 2;
}

message GossipBatchResponse {
    //The batch
    bytes content = 1;

    // The identity of the validator that requested the batch
    bytes node_id = 2;
}

message GossipBatchByBatchIdRequest {
    // The id of the batch that is being requested
    string id = 1;

    // The identity of the validator that is requesting the batch
    bytes node_id = 2;

    // A random string that provides uniqueness for requests with
    // otherwise identical fields.
    string nonce = 3;

}

message GossipBatchByTransactionIdRequest {
    // The id's of the transaction that are in the batches requested
    repeated string ids = 1;

    // The identity of the validator that is requesting the batches
    bytes node_id = 2;

    // A random string that provides uniqueness for requests with
    // otherwise identical fields.
    string nonce = 3;

}
